assets/src/js/frontend/paypal-commerce/SmartButtons.js

```diff
@@ -197,6 +204,21 @@ class SmartButtons extends PaymentMethod {
 	removeCreditCardFields() {
 		this.jQueryForm.find( 'input[name="card_name"]' ).remove();
 	}
+
+	/**
+	 * Return whether or not smart button has external or click handler.
+	 *
+	 * @since 2.9.0
+	 *
+	 * @return {boolean} Return boolean whether or not fire custom event.
+	 */
+	smartButtonHasExternalOnClickHandler() {
+		// Note: "data-customClickHandler" is only for internal use, so do not use it in production.
+		// This property allow us to submit donation form instead of processing payment withing PayPal model when click on smart button.
+		// Set it to true if you want to handle form submission on server.
+		// If donor is opted in for subscription (recurring add-on) then you do not have to overwrite this because we are handling it in recurring addon: give-recurring.js::init().
+		return 'true' === this.smartButtonContainer.getAttribute( 'data-customClickHandler' );
+	}
```

Comment:
I know this is all going to be improved later as the JS SDK is updated, so I'm going to leave this, but I wanted to point something out.

What this function is doing is inferring application state based on the DOM. That is, if the `data-customClickHandler` attribute is present on the node, then we know the smart button has an external on click handler. I do not recommend deriving state from DOM, it should be the other way around; we should have a central application state (whether it's dynamic or constant) which informs the application how to render. State, in this sense, is a higher level concept, and the DOM builds based on it. Deriving state based on DOM, means we're trying to figure out what was meant on a high level by how something was built on a low level.

Hope this helps! 😄

Reply:
@JasonTheAdams I agree with you. I did not spend too much time on this because I know that we will update it soon.